kythe/go/extractors/cmd/rust/rust_project_to_kzip.go

```diff
@@ -0,0 +1,552 @@
+type source struct {
+	IncludeDirs []string `json:"include_dirs"`
+	ExcludeDirs []string `json:"exclude_dirs"`
+}
+
+// dep is a dependency of a Crate
+type dep struct {
+	CrateId crateId `json:"crate"`
+	Name    string  `json:"name"`
+}
+
+type crate struct {
+	RootModule        string   `json:"root_module"`
+	Edition           string   `json:"edition"`
+	Deps              []dep    `json:"deps"`
+	Cfg               []string `json:"cfg"`
+	CompilerArgs      []string `json:"compiler_args"`
+	CrateId           crateId  `json:"crate_id"`
+	Label             string   `json:"label"`
+	Target            string   `json:"target"`
+	Source            source   `json:"source"`
+	IsWorkspaceMember bool     `json:"is_workspace_member"`
+}
+
+type crateId uint32
+
+// A representation of rust-project.json.
+// Note: doesn't include all fields which exist in rust-project.json, only the ones we need to convert to a kzip.
+type rustProject struct {
+	Crates []crate `json:"crates"`
+}
+
+// requiredInputsCache stores the results of file discovery operations to avoid
+// redundant filesystem walks and kzip additions. It maps include directory paths
+// (relative to the project root) to the files found within them. This is
+// particularly useful when multiple crates share common dependency source directories.
+type requiredInputsCache struct {
+	inputsCache      map[string]([]*apb.CompilationUnit_FileInput)
+	sourceFilesCache map[string][]string
+}
+
+// Check that the project root exists and is valid.
+func ensureProjectRoot(projectRoot string) {
+	if info, err := os.Stat(projectRoot); os.IsNotExist(err) {
+		log.Fatalf("Project root does not exist: %s\n", projectRoot)
+	} else if err == nil && !info.IsDir() {
+		log.Fatalf("Project root is not a directory: %s\n", projectRoot)
+	} else if err != nil {
+		log.Fatalf("Error checking project root: %v\n", err)
+	}
+}
+
+// Parse a RustProject from a path on the disk, then make all paths within that RustProject relative to a projectRoot.
+func getProjectJSON(projectJSONPath string) rustProject {
+	projectJsonFile, err := os.Open(projectJSONPath)
+	if err != nil {
+		log.Fatalf("Error opening rust-project.json: %v\n", err)
+	}
+	defer projectJsonFile.Close()
+
+	var projectJson rustProject
+	decoder := json.NewDecoder(projectJsonFile)
+	err = decoder.Decode(&projectJson)
+	if err != nil {
+		log.Fatalf("Error decoding rust-project.json: %v\n", err)
+	}
+
+	return projectJson
+}
+
+// Make all paths in includeDirs, excludeDirs, and rootModule relative to the project root
+func (project *rustProject) relativizeProjectJson(projectRoot string) {
+	for i, crate := range project.Crates {
+		for j, includeDir := range crate.Source.IncludeDirs {
+			relativized, err := removeProjectRoot(includeDir, projectRoot)
+			if err != nil {
+				fmt.Printf("Error getting relative path between %s and %s, not relativizing:%v\n", includeDir, projectRoot, err)
```

Comment:
I think `log.Printf` is a bit better: it'll print to stderr by default, support redirection, may include timestamps etc